feat(manifest): override sidecar entrypoint and cmd

buddhike opened the pull request:
This is done by following the same approach used to override default image settings.

Closes #2501

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

Commit:
feat(manifest): override sidecar entrypoint and cmd

This is done by following the same approach used to override default image settings.

Closes #2501

iamhopaul123 commented on templates/workloads/partials/cf/sidecars.yml:
```diff
@@ -18,6 +18,7 @@
 - Name: {{$sidecar.Name}}
   Image: {{$sidecar.Image}}{{if $sidecar.Essential}}
   Essential: {{$sidecar.Essential}}{{end}}{{if $sidecar.Port}}
+{{include "image-overrides" . | indent 2}}
```
Does this really work for passing `.`? I wonder if it will use the `EntryPoint` here https://github.com/aws/copilot-cli/blob/e070b5fd0844a56df59ec8ce69a95e3a3011eb3d/internal/pkg/template/workload.go#L246 instead of https://github.com/aws/copilot-cli/pull/2592/files#diff-53f2b158bad1d10014e86bdd2b5b6ccc88acb52a641ccfce4a65c9fad47517cdR94

buddhike replied:
Hey @iamhopaul123! Thanks for taking time to review this.
This is inside a `range` block that iterrates through `Sidecars` slice (https://github.com/aws/copilot-cli/pull/2592/files/4888d3522f42b8f58bb4abd00ecd084b17b7dca9#R17). Therefore `.` should refer to the current element (I double checked with go template docs here (https://pkg.go.dev/text/template#hdr-Actions).

Further to that, I did the following test with a manifest containing a different `entrypoint` config for sidecar and workload containers.

**manifest.yaml**
```
name: idgen
type: Backend Service

image:
  build: Dockerfile
  port: 8080
  depends_on:
    otel: start

entrypoint: [ gopi, idgen, --port, 8080 ]

sidecars:
  otel:
    image: otel/opentelemetry-collector-contrib:latest
    entrypoint: [otelcollector, config=/etc/otel/custom-config.yaml]
    essential: true
```
Then ran `copilot svc package --name ...` and it produced a `ContainerDefinitions` section shown below:

```
      ContainerDefinitions:
        - Name: !Ref WorkloadName
          Image: !Ref ContainerImage

          # We pipe certain environment variables directly into the task definition.
          # This lets customers have access to, for example, their LB endpoint - which they'd
          # have no way of otherwise determining.
          Environment:
          - Name: COPILOT_APPLICATION_NAME
            Value: !Sub '${AppName}'
          - Name: COPILOT_SERVICE_DISCOVERY_ENDPOINT
            Value: !Sub '${AppName}.local'
          - Name: COPILOT_ENVIRONMENT_NAME
            Value: !Sub '${EnvName}'
          - Name: COPILOT_SERVICE_NAME
            Value: !Sub '${WorkloadName}'
          - Name: OTEL_EXPORTER_OTLP_ENDPOINT
            Value: "localhost:4317"
          - Name: IDGEN_NAME
            Value:
              Fn::GetAtt: [AddonsStack, Outputs.idgenName]
          - Name: COPILOT_MOUNT_POINTS
            Value: '{"otel_config":"/etc/otel"}'

          LogConfiguration:
            LogDriver: awslogs
            Options:
              awslogs-region: !Ref AWS::Region
              awslogs-group: !Ref LogGroup
              awslogs-stream-prefix: copilot

          EntryPoint:
          - gopi
          - idgen
          - --port
          - 8080
          MountPoints:
            - ContainerPath: '/etc/otel'
              ReadOnly: true
              SourceVolume: otel_config
          DependsOn:
            - Condition: START
              ContainerName: otel
          PortMappings: !If [ExposePort, [{ContainerPort: !Ref ContainerPort}], !Ref "AWS::NoValue"]

        - Name: otel
          Image: otel/opentelemetry-collector-contrib:latest
          Essential: true

          EntryPoint:
          - otelcollector
          - config=/etc/otel/custom-config.yaml
          PortMappings:
            - ContainerPort: 4317
          LogConfiguration:
            LogDriver: awslogs
            Options:
              awslogs-region: !Ref AWS::Region
              awslogs-group: !Ref LogGroup
              awslogs-stream-prefix: copilot
          MountPoints:
            - SourceVolume: otel_config
              ReadOnly: true
              ContainerPath: '/etc/otel'
```

Hopefully, this makes sense. Please let me know if I'm missing something or if you'd like to include another test scenario 🙏🏾.

iamhopaul123 replied:
Gotcha. That looks good to me! Thank you for the manual test.

iamhopaul123 reviewed (commented):
Thank you so much for contributing to this! Manifest -> stack looks good but i do have a question from stack -> CFN template. Also could you update the doc for this new feature as well? Thank you!

buddhike commented:
Sure thing, I will update the doc as well 😊

Commit:
docs: sidecar entrypoint and command overrides

All sidecar options including the new entrypoint and command overrides are moved to a new partial file so that it can be included  anywhere in docs.

buddhike commented:
Doc is updated 🍻

iamhopaul123 reviewed (approved):
Ship it 🚀

Lou1415926 reviewed (approved):
Thank you so so much for helping us update the doc and the awesome code! 🚢  it!

mergify[bot] committed:
Merge branch 'mainline' into ft/sidecar-options
